18723 - pass some of advisor config via --scheduler-config

linsun opened the pull request:
@jdef -- reworked the patch based on your late last night's comment.  thank you for your help!

linsun committed:
#18723 - Ability to pass cadvisor args to mesos executor

linsun committed:
additional fix for 18723 - use the new flags when launching the executor

linsun committed:
additional fix for 18723 - clean up

linsun commented:
I have signed the IBM corporate CLA -- My ibm email is linsun@us.ibm.com

MikeSpreitzer commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -351,7 +361,7 @@ func (s *SchedulerServer) serveFrameworkArtifactWithFilename(path string, filena
 
 func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.ExecutorInfo, error) {
-	ci := &mesos.CommandInfo{
+		ci := &mesos.CommandInfo{
```
Lin, did your `git commit` trigger a go syntax checker?  If not then you need to upgrade your methodology.

MikeSpreitzer replied:
@linsun: I am referring specifically to http://kubernetes.io/v1.1/docs/devel/development.html#hooks

linsun replied:
@MikeSpreitzer I just added the hooks - thanks!   Looking at the kubernetes master tree, the code was not correctly formatted:  https://github.com/kubernetes/kubernetes/blob/master/contrib/mesos/pkg/scheduler/service/service.go -- line 354, so the patch is ok.

jdef replied:
this still isn't properly formatted

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -333,6 +336,10 @@ func (ms *MinionServer) AddExecutorFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&ms.logVerbosity, "v", ms.logVerbosity, "log level for V logs")
+	
+	// hacks to forward selected cadvisor flags to the executor
+	fs.StringVar(&ms.housekeepingInterval, "housekeeping_interval", ms.housekeepingInterval, "Interval between container housekeepings")
```
from our build system:
```
[12:26:18][Step 6/6] # k8s.io/kubernetes/contrib/mesos/pkg/minion
[12:26:18][Step 6/6] _output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/minion/server.go:341: cannot use &ms.housekeepingInterval (type *time.Duration) as type *string in argument to fs.StringVar
[12:26:18][Step 6/6] _output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/minion/server.go:341: cannot use ms.housekeepingInterval (type time.Duration) as type string in argument to fs.StringVar
[12:26:18][Step 6/6] _output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/minion/server.go:342: cannot use &ms.globalHousekeepingInterval (type *time.Duration) as type *string in argument to fs.StringVar
[12:26:18][Step 6/6] _output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/minion/server.go:342: cannot use ms.globalHousekeepingInterval (type time.Duration) as type string in argument to fs.StringVar
```

linsun replied:
@jdef Thanks - will rework.   How can I trigger the smoke test myself locally?  I have been trying to run:
hack/build-go.sh
hack/test-go.sh -- is this same as the smoke test?

jdef replied:
if you have the prerequisites installed on your dev machine (docker,
docker-compose, etc) you can run `./contrib/mesos/ci/test-smoke.sh`

On Fri, Jan 29, 2016 at 1:23 PM, Lin Sun <notifications@github.com> wrote:

> In contrib/mesos/pkg/minion/server.go
> <https://github.com/kubernetes/kubernetes/pull/20331#discussion_r51295425>
> :
>
> > @@ -333,6 +336,10 @@ func (ms *MinionServer) AddExecutorFlags(fs *pflag.FlagSet) {
> >
> >  	// hack to forward log verbosity flag to the executor
> >  	fs.Int32Var(&ms.logVerbosity, "v", ms.logVerbosity, "log level for V logs")
> > +	
> > +	// hacks to forward selected cadvisor flags to the executor
> > +	fs.StringVar(&ms.housekeepingInterval, "housekeeping_interval", ms.housekeepingInterval, "Interval between container housekeepings")
>
> @jdef <https://github.com/jdef> Thanks - will rework. How can I trigger
> the smoke test myself locally? I have been trying to run:
> hack/build-go.sh
> hack/test-go.sh -- is this same as the smoke test?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331/files#r51295425>.
>

linsun replied:
@jdef I'm having trouble to run test-smoke.sh from my Mac.  I'll try attempt to run it from a Linux box.  Is this a known issue?

Another q, I got this error from the pre-commit hook:
Are you certain this flag should not have been declared with an - instead?
global_housekeeping_interval
housekeeping_interval

Should i change them to use global-housekeeping-interval and housekeeping-interval instead?  or shall i just update the excluded-flags.txt?  Thank you!

jdef commented:
looking better. good rule of thumb: pushed code should compile and pass unit tests.

jdef commented:
@ixdy not sure what's going on with the CLA bot here

linsun committed:
address smoke test issue from 18723

jdef commented:
try naming the flags with `-` and see if the problem goes away. i think
there's some flag munging magic happening somewhere..

On Fri, Jan 29, 2016 at 2:49 PM, googlebot <notifications@github.com> wrote:

> CLAs look good, thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331#issuecomment-176937567>
> .
>

jdef commented:
check out the docs here:
https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/mesos-docker.md

On Fri, Jan 29, 2016 at 2:57 PM, James DeFelice <james@mesosphere.io> wrote:

> try naming the flags with `-` and see if the problem goes away. i think
> there's some flag munging magic happening somewhere..
>
> On Fri, Jan 29, 2016 at 2:49 PM, googlebot <notifications@github.com>
> wrote:
>
>> CLAs look good, thanks!
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/20331#issuecomment-176937567>
>> .
>>
>
>

linsun committed:
change to use - instead of _

linsun commented:
@jdef sorry for the delay.  I've reworked for the "_" issue.   It took me quite long to get through running the smoke-test locally on my MacOS.  I ran into a few issues (opened as JIRA) that I was able to resolve, but a few I could not.  So at the end, I was able to run the test on a ubuntu VM.   Here is the output, looks good:

root@linsun3:~/kubernetes# ./contrib/mesos/ci/test-smoke.sh -v=2
Cleaning work dir
/root/tmp/kubernetes
Detecting docker client
/usr/bin/docker
+ make clean all
build/make-clean.sh
+++ [0202 20:28:33] Verifying Prerequisites....
!!! [0202 20:28:34] Build image not built.  Cannot clean via docker build image.
+++ [0202 20:28:34] Removing data container
+++ [0202 20:28:34] Cleaning out local _output directory
+++ [0202 20:28:34] Deleting docker image kube-build:build-f76adbc52b
+++ [0202 20:28:34] Cleaning all other untagged docker images
Deleted: ba1677efed510dec66c1d742103f76b43bb418ce244ab5ba6c80dba6d66274f9
Deleted: 875d004ee6597a34b5525b183959e28df22ba48dcc01b43f0c18fdb1d3a8eb03
Deleted: 99153bb5b2cf27c8b51485e452071e37ddb26a2d5b789c0845c72a607a3d20fc
Deleted: 6d68ea22f393ee43b212907b8968b58b4d5f7a4ab4de4c0ca49a9bb833f7a93e
Deleted: 8ad6758385b1a2ff8f284ac5914f48501cba240a73956462ccc89df96c1cf5ac
Deleted: b3685df6252ebbcd67416d78e22986e03953368fe053de226472afb7631555f5
Deleted: 4cd6a96cfe656498bbf32058ffaaa69ce4ae69d896b4629757610a4bc43db3aa
Deleted: a93d2ef00ae9abcd38002fa566201d335771a1e7566b5c51dbc2b9bf3fb7bd5f
Deleted: 2d5960ddd81d6a1711d6bb8b1f68c54e1377c5fb15df9b8e865cdeb23cb696e7
Deleted: 38741b71be8687ab24f412c003a6d30753c497c078f8932b361142cb8756c745
Deleted: fb0ba56c2d7844b28464b530cf61e5bdf0fdaf0d65c021b43023151e93df1651
Deleted: 764bc9c15e70e45bf163710f8b71cbabcfe15e661426c4c673613101c203996d
Deleted: 414bfa7216615ad4ee873001c7ff4379426e97b47813f3c149548cd64779f2af
Deleted: 79cd02447b4db7b4e68f784410fb0452ce886cf773ba5f288be4960c72a39614
Deleted: 63d1729bdcc92e4ff84aa55026ca8f14a85e968c225df09fa732620774f3ee5c
Deleted: 43f00c01b3a28a220bbc8579cdae9aa462f966b17c59d8d6b46541d121c69ab4
Deleted: ce6fcdb5c008215409b7944f393f82a9e635aeec94a38a1ee2960cac68b86783
Deleted: 91f2eb3bc0392f5b43c722cfd0fc99e0b49c44d94ff221fd040d8d79bece8874
Deleted: 0d376a88d592a91381c4e09fd93b63f57bae20f83678d494a2ca355e596de6f4
Deleted: 96078bbbd1133a7af654b85a2219273eeac3a85e4e60e45b9fdf4fd048ad8106
Deleted: 123bb2b94eceaec275718da18e7f5bcd992e699ba62142754ee7b0826b6cb135
Deleted: afc814718fc08e54a3ee3220fc52ea30301ca308281154192178fffe84049de8
Deleted: 96da4bbafa3a05bf063d3f4ed4b52380c4b9e5351ab8aa04214817e7b0485688
Deleted: 4972f75c2dee6b440555c373e61d6f1fd48bbb105b6f97fa96c227870de04c7a
Deleted: e94a21a29063f83eed34bd448426cd78dd78dc4b66509a3c429027ff958d1d21
Deleted: 0f4e2d3ead09cf0fde830ef214bcca12c94771e6dbe10dfbba5b422ed0c98d3c
Deleted: 798e241a59714e76809a59ca0d2e5a0efbc41d99917121bf65e445daedd7afef
Deleted: 73051696f5f0e8df70eb19f3182f1b983b7c2652121c14f044592bd5d92c955c
Deleted: 5c099df10e506f43adf9724c50ee051b64fd2d4fa40bdb41dcaa5ba91c3c4615
rm -rf _output
rm -rf Godeps/_workspace/pkg
hack/build-go.sh 
+++ [0202 20:28:36] Building go targets for linux/amd64:
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubemark
    cmd/hyperkube
    cmd/linkcheck
    plugin/cmd/kube-scheduler
    contrib/mesos/cmd/k8sm-scheduler
    contrib/mesos/cmd/k8sm-executor
    contrib/mesos/cmd/k8sm-controller-manager
    contrib/mesos/cmd/km
    cmd/kubectl
    cmd/integration
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/mungedocs
    cmd/genbashcomp
    cmd/genconversion
    cmd/gendeepcopy
    cmd/genswaggertypedocs
    examples/k8petstore/web-server/src
    github.com/onsi/ginkgo/ginkgo
    test/e2e/e2e.test
+++ [0202 20:30:24] Placing binaries
+ trap 'timeout 5m ./cluster/kube-down.sh' EXIT
+ ./cluster/kube-down.sh
Bringing down cluster using provider: mesos/docker
Verifying required commands
Stopping mesos/docker cluster
No stopped containers
Done
+ ./cluster/kube-up.sh
... Starting cluster using provider: mesos/docker
... calling verify-prereqs
Verifying required commands
... calling kube-up
Creating Mesos Work Dir: /root/tmp/kubernetes/mesosslave
Pulling Docker images
Building Docker images
Installing nsenter to /target
Installing docker-enter to /target
Installing importenv to /target
Workspace created: /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-workspace-X0c6MM
Copying files to workspace
./
./bin/
./bin/socat
./bin/nsenter
/go/src/github.com/GoogleCloudPlatform/kubernetes
Building docker image mesosphere/kubernetes-mesos:latest
+ docker build -t mesosphere/kubernetes-mesos:latest .
Sending build context to Docker daemon 80.02 MB
Step 1 : FROM ubuntu:14.04.3
 ---> 6cc0fc2a5ee3
Step 2 : MAINTAINER Mesosphere <support@mesosphere.io>
 ---> Using cache
 ---> 31f6beb01e53
Step 3 : RUN locale-gen en_US.UTF-8
 ---> Using cache
 ---> 7d1b96880cef
Step 4 : RUN dpkg-reconfigure locales
 ---> Using cache
 ---> 670d08e556e3
Step 5 : ENV LANG en_US.UTF-8
 ---> Using cache
 ---> 67705219cdfb
Step 6 : ENV LC_ALL en_US.UTF-8
 ---> Using cache
 ---> 006885c46ca3
Step 7 : RUN apt-get update -qq &&     DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -qqy         ca-certificates         wget         curl         &&     apt-get clean
 ---> Using cache
 ---> 68c9fd3c883d
Step 8 : RUN curl -o- https://raw.githubusercontent.com/karlkfi/resolveip/v1.0.2/install.sh | bash
 ---> Using cache
 ---> ac238fa132af
Step 9 : COPY ./bin/* /usr/local/bin/
 ---> e5494e320db5
Removing intermediate container 012b6da2a597
Step 10 : COPY ./opt/* /opt/
 ---> 168c719e9b73
Removing intermediate container 2b139b79e543
Successfully built 168c719e9b73
+ set +o xtrace
Built docker image mesosphere/kubernetes-mesos:latest
Workspace deleted: /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-workspace-X0c6MM
Workspace created: /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv
+ echo 'Copying files to workspace'
Copying files to workspace
+ mkdir -p /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv/bin
+ cp -a /go/src/github.com/GoogleCloudPlatform/kubernetes/cluster/mesos/docker/common/bin/await-file /go/src/github.com/GoogleCloudPlatform/kubernetes/cluster/mesos/docker/common/bin/await-health-check /go/src/github.com/GoogleCloudPlatform/kubernetes/cluster/mesos/docker/common/bin/health-check /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv/bin/
+ cp -a /go/src/github.com/GoogleCloudPlatform/kubernetes/cluster/mesos/docker/test/bin/install-etcd.sh /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv/bin/
+ cp -a /go/src/github.com/GoogleCloudPlatform/kubernetes/cluster/mesos/docker/test/Dockerfile /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv/
+ cd /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv
+ echo 'Building docker image mesosphere/kubernetes-mesos-test:latest'
Building docker image mesosphere/kubernetes-mesos-test:latest
+ set -o xtrace
+ docker build -t mesosphere/kubernetes-mesos-test:latest .
Sending build context to Docker daemon 11.78 kB
Step 1 : FROM golang:1.4.3
 ---> da0462d15648
Step 2 : MAINTAINER Mesosphere <support@mesosphere.io>
 ---> Using cache
 ---> 97bfca9c13d2
Step 3 : RUN apt-get update -qq &&     DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -qqy         wget         curl         g++         make         mercurial         git         rsync         patch         python         python-pip         apt-transport-https         &&     apt-get clean
 ---> Using cache
 ---> 5b0a4bbee3ce
Step 4 : RUN apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D &&     mkdir -p /etc/apt/sources.list.d &&     echo deb https://apt.dockerproject.org/repo ubuntu-trusty main > /etc/apt/sources.list.d/docker.list &&     apt-get update -qq &&     DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -qqy         docker-engine=1.8.1-0~trusty         &&     apt-get clean
 ---> Using cache
 ---> d5beb7eb0c1c
Step 5 : RUN pip install -U docker-compose==1.5.0
 ---> Using cache
 ---> f3bbe87aeabd
Step 6 : RUN go get github.com/tools/godep
 ---> Using cache
 ---> 77a1a8fe19b0
Step 7 : RUN mkdir -p /go/src/github.com/GoogleCloudPlatform/kubernetes
 ---> Using cache
 ---> 5987e2c4ea4c
Step 8 : WORKDIR /go/src/github.com/GoogleCloudPlatform/kubernetes
 ---> Using cache
 ---> ef9b243ab28e
Step 9 : COPY ./bin/* /usr/local/bin/
 ---> Using cache
 ---> 3f11a691b898
Step 10 : RUN install-etcd.sh
 ---> Using cache
 ---> 42ef3e4929e5
Step 11 : ENTRYPOINT bash
 ---> Using cache
 ---> a85ff5c12886
Successfully built a85ff5c12886
+ set +o xtrace
Built docker image mesosphere/kubernetes-mesos-test:latest
Workspace deleted: /go/src/github.com/GoogleCloudPlatform/kubernetes/k8sm-test-workspace-rWY0iv
Creating Auth Dir: /root/tmp/kubernetes/auth
Creating Certificate Authority
Certificate Authority Key: /root/tmp/kubernetes/auth/root-ca.key
Certificate Authority Cert: /root/tmp/kubernetes/auth/root-ca.crt
Creating Service Account RSA Key
Service Account Key: /root/tmp/kubernetes/auth/service-accounts.key
Creating User Accounts
Token Users: /root/tmp/kubernetes/auth/token-users
Basic-Auth Users: /root/tmp/kubernetes/auth/basic-users
Starting mesos/docker cluster
Creating docker_ambassador_1
Creating docker_etcd_1
Creating docker_mesosmaster1_1
Creating docker_apiserver_1
Creating docker_keygen_1
Creating docker_controller_1
Creating docker_mesosslave_1
Creating docker_scheduler_1
Scaling mesos/docker cluster to 2 slaves
Creating and starting 2 ... done
Waiting (up to 180s) for http://apiserver:8888/healthz to be healthy
Health check of http://apiserver:8888/healthz succeeded!
KUBE_MASTER_IP: 172.17.0.5:6443
KUBE_NODE_IP_ADDRESSES: [172.17.0.10 172.17.0.8]
cluster "mesos/docker" set.
context "mesos/docker" set.
user "cluster-admin" set.
switched to context "mesos/docker".
Wrote config for mesos/docker to /root/.kube/config
Deploying Addons
namespace "kube-system" created
namespace "static-pods" created
Deploying DNS Addon
Workspace created: /tmp/kube-dns.lmVBfI
replicationcontroller "kube-dns-v10" created
service "kube-dns" created
Workspace deleted: /tmp/kube-dns.lmVBfI
Deploying UI Addon
replicationcontroller "kube-ui-v4" created
service "kube-ui" created
kube-dns: ....................................................................................................................................................................................Pending
Dumping logs to '/root/tmp/kubernetes/log'
+ timeout 5m ./cluster/kube-down.sh
Bringing down cluster using provider: mesos/docker
Verifying required commands
Stopping mesos/docker cluster
Killing docker_mesosslave_2 ... done
Killing docker_scheduler_1 ... done
Killing docker_mesosslave_1 ... done
Killing docker_controller_1 ... done
Killing docker_apiserver_1 ... done
Killing docker_mesosmaster1_1 ... done
Killing docker_etcd_1 ... done
Killing docker_ambassador_1 ... done
Going to remove docker_mesosslave_2, docker_scheduler_1, docker_mesosslave_1, docker_controller_1, docker_keygen_1, docker_apiserver_1, docker_mesosmaster1_1, docker_etcd_1, docker_ambassador_1
Removing docker_mesosslave_2 ... done
Removing docker_scheduler_1 ... done
Removing docker_mesosslave_1 ... done
Removing docker_controller_1 ... done
Removing docker_keygen_1 ... done
Removing docker_apiserver_1 ... done
Removing docker_mesosmaster1_1 ... done
Removing docker_etcd_1 ... done
Removing docker_ambassador_1 ... done
Done

linsun commented:
I'm checking on the above failure from "Travis CI build failed".

linsun committed:
gofmt the file & one more spot for the - change

linsun commented:
@jdef looking at the latest CI test result, it seems cadvisor is expecting "_" like housekeeping_interval, after I change it to housekeeping-interval.  Any advice?

jdef commented:
the current combination of cadvisor+kubelet is somewhat dirty: cadvisor is
declaring global flags (instead of using flag sets) and kubelet is picking
these up via it's cadvisor dependencies. let me re-review the patch and
make another recommendation.

On Tue, Feb 2, 2016 at 4:35 PM, Lin Sun <notifications@github.com> wrote:

> @jdef <https://github.com/jdef> looking at the latest CI test result, it
> seems cadvisor is expecting "_" like housekeeping_interval, after I change
> it to housekeeping-interval. Any advice?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331#issuecomment-178834843>
> .
>

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -74,6 +75,8 @@ type MinionServer struct {
 	proxyMode                      string
 	conntrackMax                   int
 	conntrackTCPTimeoutEstablished int
+	globalHousekeepingInterval     time.Duration
```
provide default values for these in the MinionServer initializer

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -333,6 +336,10 @@ func (ms *MinionServer) AddExecutorFlags(fs *pflag.FlagSet) {
 	// hack to forward log verbosity flag to the executor
 	fs.Int32Var(&ms.logVerbosity, "v", ms.logVerbosity, "log level for V logs")
+
+	// hacks to forward selected cadvisor flags to the executor
```
move these to AddMinionFlags

jdef replied:
... and in `launchExecutorServer` add some logic to add arguments `--global_housekeeping_interval=%v` and `--housekeeping_interval=%v` using the values from the minion struct.

jdef commented:
OK I've left some additional feedback that effectively makes the "-" flags
scheduler/minion flags and then the minion would translate those into "_"
flags when launching the kubelet-executor. hopefully we can clean that up
when the kubelet/cadvisor integration simplifies and we're not forced to
swallow "_" global flags.

On Tue, Feb 2, 2016 at 6:59 PM, James DeFelice <james@mesosphere.io> wrote:

> the current combination of cadvisor+kubelet is somewhat dirty: cadvisor is
> declaring global flags (instead of using flag sets) and kubelet is picking
> these up via it's cadvisor dependencies. let me re-review the patch and
> make another recommendation.
>
> On Tue, Feb 2, 2016 at 4:35 PM, Lin Sun <notifications@github.com> wrote:
>
>> @jdef <https://github.com/jdef> looking at the latest CI test result, it
>> seems cadvisor is expecting "_" like housekeeping_interval, after I change
>> it to housekeeping-interval. Any advice?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/20331#issuecomment-178834843>
>> .
>>
>
>

linsun committed:
adding output from running verify-flags-underscore.py

linsun commented:
@jdef thank you much for the feedback. I've reworked, and will watch out for the CI results.

linsun commented:
Continuous CI looks green.  Github CI tests are not green -- is there anything I need to do on my end to make this green?

linsun commented:
Just saw the comment from k8s-teamcity-mesosphere... not sure why I didn't get a notification email.  Will work on that.

jdef commented:
are you setting an envvar `KUBERNETES_CONTRIB=mesos` when you run `make`? if not, then it's probably not compiling the k8s-mesos code on your machine.

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -174,6 +179,9 @@ func (ms *MinionServer) launchExecutorServer() <-chan struct{} {
 		executorArgs = append(executorArgs, "--cgroup-root="+ms.cgroupRoot)
 	}
 
+	executorArgs = append(executorArgs, "--housekeeping_interval="+ms.housekeepingInterval)
```
s/)/.String())/

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -174,6 +179,9 @@ func (ms *MinionServer) launchExecutorServer() <-chan struct{} {
 	}
 
+	executorArgs = append(executorArgs, "--housekeeping_interval="+ms.housekeepingInterval)
+	executorArgs = append(executorArgs, "--global_housekeeping_interval="+ms.globalHousekeepingInterval)
```
ditto

linsun committed:
fix compile error in smoke_test

linsun commented:
Thank you much @jdef!  Yes, I forgot to set the env var.   I have run make with the env var, plus the smoke test locally.  I hope everything is good now.

jdef commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -225,6 +227,10 @@ func NewSchedulerServer() *SchedulerServer {
 		conntrackMax: 0,
+
+		// cadvisor housekeeping related config
+		housekeepingInterval:       1 * time.Second,
```
these look different than the defaults for the minion. why? the cadvisor default for `housekeeping_interval` is `10s` -- any reason to use something different here? if so, can you please comment as to why?

linsun replied:
Hmm, I was reading it from cadvisor/docs/runtime_options.md, it had:

--housekeeping_interval=1s: Interval between container housekeepings

jdef replied:
when I compiled your branch and looked at the flags of `km scheduler` I see
this:
```
      --housekeeping-interval=1s: Interval between container housekeepings
      --housekeeping_interval=10s: Interval between container housekeepings
```

the 1s belonged to the flag that you created, the 10s belonged to the
cadvisor global flag.


On Thu, Feb 4, 2016 at 4:19 PM, Lin Sun <notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/20331#discussion_r51938799>
> :
>
> > @@ -225,6 +227,10 @@ func NewSchedulerServer() *SchedulerServer {
> >  		// when kube-proxy is running in a non-root netns (init_net); setting this to a non-zero value will
> >  		// impact connection tracking for the entire host on which kube-proxy is running. xref (k8s#19182)
> >  		conntrackMax: 0,
> > +
> > +		// cadvisor housekeeping related config
> > +		housekeepingInterval:       1 * time.Second,
>
> Hmm, I was reading it from cadvisor/docs/runtime_options.md, it had:
>
> --housekeeping_interval=1s: Interval between container housekeepings
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331/files#r51938799>.
>

linsun replied:
Thank you @jdef - I'll try look at this tonight to see why.

linsun replied:
I am very puzzled on where the housekeeping_interval=10s is specified...  I searched the cadvisor code, only see 1s is specified.   it seems we should remove it from 'km scheduler --help', because we don't support passing it via the housekeeping_interval with "_" in it (only dash is supported).

jdef replied:
i don't think there's a way to remove it from `km scheduler --help`. km is
a "hyperkube" binary, meaning that it links together several different
binaries into one. since one of those binaries is the kubelet-executor
(which links against cadvisor) and cadvisor is a bad actor (by defining
global flags), we end up with the cadvisor flags showing up alongside the
scheduler flags

this feels somewhat unsatisfying. need to think about a better way to
handle these flags. having both the new and the old flags show up is very
confusing -- even if it does satisfy the build/verification scripts.

thanks for your patience. will need to think on this.



On Thu, Feb 4, 2016 at 10:13 PM, Lin Sun <notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/20331#discussion_r51972674>
> :
>
> > @@ -225,6 +227,10 @@ func NewSchedulerServer() *SchedulerServer {
> >  		// when kube-proxy is running in a non-root netns (init_net); setting this to a non-zero value will
> >  		// impact connection tracking for the entire host on which kube-proxy is running. xref (k8s#19182)
> >  		conntrackMax: 0,
> > +
> > +		// cadvisor housekeeping related config
> > +		housekeepingInterval:       1 * time.Second,
>
> I am very puzzled on where the housekeeping_interval=10s is specified... I
> searched the cadvisor code, only see 1s is specified. it seems we should
> remove it from 'km scheduler --help', because we don't support passing it
> via the housekeeping_interval with "_" in it (only dash is supported).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331/files#r51972674>.
>

linsun committed:
update exception file

linsun commented:
No disagreement -- having both is confusing, even tho the _ ones won't work....  let me know other thoughts you may have.  thanks @jdef !

linsun commented:
maybe one way is to change back to use _ and add housekeeping_interval to the exception list...?

jdef commented:
Perhaps we try this approach? Since the `_` flags are present in every `km` binary I think we should write a helper (perhaps in `contrib/mesos/pkg/cadvisor`) to simplify access to the global flags of cadvisor, along the lines of what the example scheduler from [mesos-go](https://github.com/mesos/mesos-go/blob/master/examples/scheduler/main.go#L235) does.

For example:

```go
func HousekeepingInterval(defaultValue time.Duration) (v time.Duration) {
	v = defaultValue
	if f := flag.Lookup("housekeeping_interval"); f != nil && f.Value != nil {
		if vstr := f.Value.String(); vstr != "" {
			v = // ... attempt to parse it here
		}
	}
	return
}
```

* The scheduler would invoke the above func and pass `housekeeping_interval` flag to the executor, always
* The minion would invoke the above func and pass `housekeeping_interval` flag to the kubelet-executor process
* The scheduler and minion `xxxServer` structs don't require any additional fields
* No additional flags are required at all by k8sm
* Should add a nice big TODO near these funcs in the new `contrib/mesos/pkg/cadvisor` package that says how the code should die once cadvisor is more properly integrated into the kubelet
  * we'll know exactly when this happens because smoke tests will consistently fail since the `_` flags will have disappeared from the km binary

linsun commented:
@jdef, thank you for your comment!  I'm looking into that and will provide update.

linsun committed:
resolve conflict after merging latest k8s

jdef commented:
were there any additional questions/concerns with the approach I suggested?

linsun commented:
Sorry for the delayed reply!  My priority has changed, and I hope to look at this soon!

linsun committed:
resolving merge conflict

linsun commented:
Sorry for the super long delay - just started to look at this.  @jdef, I don't fully understand your comment on Feb5.  Are you saying creating a new pkg contrib/mesos/pkg/cadvisor and put the helper go class in there?  If so, how does this helper method getting the flags passed from user from the Kubernetes scheduler?

Also, why cannot I simply not use -, e.g. just use housekeeping_interval and remove the need for housekeeping-interval?

linsun committed:
change to use _ instead of -

linsun committed:
results of running gofmt and adding to exclude list

linsun commented:
@jdef with my latest change, the km scheduler output looks ok, i.e. I no longer get 2 entries for each.

      --global_housekeeping_interval=1m0s: Interval between container global housekeepings
      --housekeeping_interval=1s: Interval between container housekeepings

I'm checking on my CLA, and check if I can test this.  Thanks.

jdef commented:
sorry it's taken me a while to get back to this. thanks for revising this PR. i think there's a way forward without adding flags to the k8sm servers the "normal" way, but instead to just leverage the fact that the cadvisor flags are global and use that to more simply forward the values from process to process. this is what i was trying to say back in Feb (from your response it seems that my writing did not properly convey the idea).

the cadvisor flags are being refactored and will, at some point, no longer be global. we're still waiting to see what form that will take. in the meantime, i'm trying to avoid changes to the scheduler and minion flag sets for this. 

see https://github.com/kubernetes/kubernetes/pull/22974

linsun commented:
@jdef thank you for the code change and other pointers!  Sorry I didn't quite understand how the util class can get the two flags earlier.  I'm going to close this pull request and give your changes a try!

jdef commented:
great - let me know if it works for you. if so, maybe we can squeeze this
changeset into the 1.2 release

On Tue, Mar 15, 2016 at 1:54 PM, Lin Sun <notifications@github.com> wrote:

> @jdef <https://github.com/jdef> thank you for the code change and other
> pointers! Sorry I didn't quite understand how the util class can get the
> two flags earlier. I'm going to close this pull request and give your
> changes a try!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20331#issuecomment-196946383>
>

linsun commented:
This issue can be closed now, given @jdef has merged in this commit (thanks much!):  https://github.com/kubernetes/kubernetes/pull/22974
